dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java

```diff
@@ -410,20 +430,161 @@ public Workflow refresh(@ApiParam(hidden = true) @Auth User user,
             @ApiParam(value = "repository path", required = true) @PathParam("repository") String repository,
             @ApiParam(value = "Git reference for new GitHub tag", required = true) @QueryParam("gitReference") String gitReference,
             @ApiParam(value = "This is here to appease Swagger. It requires PUT methods to have a body, even if it is empty. Please leave it empty.") String emptyBody) {
+        // Call common upsert code
+        String dockstoreWorkflowPath = upsertVersionHelper(repository, gitReference, user, WorkflowMode.FULL, null);
+
+        return findAllWorkflowsByPath(dockstoreWorkflowPath, WorkflowMode.FULL);
+    }
+
+    /**
+     * Finds all workflows from a general Dockstore path that are of type FULL
+     * @param dockstoreWorkflowPath Dockstore path (ex. github.com/dockstore/dockstore-ui2)
+     * @return List of FULL workflows with the given Dockstore path
+     */
+    private List<Workflow> findAllWorkflowsByPath(String dockstoreWorkflowPath, WorkflowMode workflowMode) {
+        return workflowDAO.findAllByPath(dockstoreWorkflowPath, false)
+                .stream()
+                .filter(workflow ->
+                        workflow.getMode() == workflowMode)
+                .collect(Collectors.toList());
+    }
+
+    @PUT
+    @Path("/path/service/{repository}")
+    @Timed
+    @UnitOfWork
+    @RolesAllowed({ "curator", "admin" })
+    @ApiOperation(value = "Create a service for the given repository (ex. dockstore/dockstore-ui2).", notes = "To be called by a lambda function.", authorizations = {
+            @Authorization(value = JWT_SECURITY_DEFINITION_NAME) }, response = Service.class)
+    public Service addService(@ApiParam(hidden = true) @Auth User user,
+            @ApiParam(value = "Repository path", required = true) @PathParam("repository") String repository,
+            @ApiParam(value = "Name of user on GitHub", required = true) @QueryParam("username") String username,
+            @ApiParam(value = "GitHub installation ID", required = true) @QueryParam("installationId") String installationId,
+            @ApiParam(value = "This is here to appease Swagger. It requires PUT methods to have a body, even if it is empty. Please leave it empty.") String emptyBody) {
+        // Check for duplicates (currently workflows and services share paths)
+        String servicePath = "github.com/" + repository;
+        Workflow duplicate = workflowDAO.findByPath(servicePath, false);
 
+        if (duplicate != null) {
+            String msg = "Service with the path " + servicePath + " already exists.";
+            LOG.info(msg);
+            throw new CustomWebApplicationException(msg, HttpStatus.SC_BAD_REQUEST);
+        }
+
+        // Retrieve the user who triggered the call
+        User sendingUser = findUserByGitHubUsername(username);
```

Comment:
What if user doesn't exist? I guess that should never happen.... Although it makes me think of a broader question -- if you delete your Dockstore user account, do the apps get uninstalled?

Reply:
For the latter broader question, my understanding is that app installs aren't really associated with the user who installed them. So deleting your (GitHub, not Dockstore account) won't affect the GitHub app.

Reply:
I'm  wondering about the case when a user deletes a Dockstore account. If the web hook still fires for those repos (?), then the above line will return null.

Reply:
Ah I see, yes. The user that triggered on GitHub might not be a Dockstore user anyway. Line 483 should probably be tolerant of an unknown/null user. (i.e. I'd expect the version to still be added, but no user added)

Reply:
@coverbeck @denis-yuen  this endpoint is only called for adding a service, and not for adding versions. Only when we are adding a service do we check for a user being associated with the repo. When we are adding a version to a service, we don't care about the user.

Though now I just realised that this endpoint could be called for an existing service but a new user has installed, so it should add them as a user without adding a new service.

Reply:
Makes sense, voted both ways for the last bit. (I don't think two users can install the same github app to the same repo, but please double check that)

Reply:
You are correct, one install per repository.